Simplified(?) egg cracking 🥚 🔨

DoctorVanGogh opened the pull request:
I want to be able to crack eggs into cauldrons. 

But to enable that I need to understand how the egg-cracking works.

So this is me refactoring the egg cracking logic.

I *think* this is way more maintainable. Eliminated several deeply branching repetitive (but not identical!) blocks and replaced them with a single custom method.
This _should_ be easier to follow.

Also added the ability to *target* which ground receptacle to crack into (before the game always picked "first available". Now you can choose which one).

![image](https://user-images.githubusercontent.com/324067/217328670-1cec9c6b-e796-488c-8ddd-c66a5e3afd49.png)

Other subtle change: Made the "no cracking" behaviour when target container is full consistent between bucket (or similar) targets & bowls. Now not just the bucket stops waste, the bowl does too. 
You can *also* successfully crack into jugs. As far as vanilla is concerned jugs are valid squeeze targets (because they have an open top). Pinging @MAGGen-hub here, since you merged a jug related change recently and I don't 100% understand what that did. Was cracking into jugs working incorrectly? Should it not have worked?

DoctorVanGogh commented:
Tested with buckets, fired & unfired bowls & jugs. Both in 1 or 4 per block configurations. Everything works as expected.

Can add to fired versions until full. Can't add to unfired versions. Can't add different result than what's already in there.

Heck, can even add to condensers, Because that's what the rolling interaction tooltip said 😁 

Did **not** test for cross mixing with different species eggs (since I have no additional egg species installed). But should work since the existing check for that is still in there (btw. you can currently mix `foo` egg whites & `bar` egg whites just fine).

DoctorVanGogh commented:
Okay... I've come up with a potentially 'simpler' nested branching for the crackability decision but I dont really have a preference on either.

### Version A (current):
```cs
            var liquidOptions = GetLiquidOptions(blockAccessor, blockSel);

            string eggType = slot.Itemstack.Collectible.FirstCodePart(0);      //grabs currently held item's code
            string eggVariant = slot.Itemstack.Collectible.FirstCodePart(1);   //grabs 1st variant in currently held item

            bool canCrack = (liquidOptions, liquidOptions?.ExistingStack?.Collectible?.FirstCodePart(0), liquidOptions?.ExistingStack?.Collectible?.FirstCodePart(1)) switch {
                (null, _, _)                                => false,                                                                               // cant add liquid
                ( { ExistingStack: null}, _, _)             => true,                                                                                // empty target stack - can always squeeze (CanAddLiquid already checks for CanSqueeze)
                (_, "eggyolkfullportion", var yolk)         => byEntity.Controls.Sprint && IsCrackableEggType(eggType) && yolk == eggVariant,       // liquid egg in container, need to be full cracking, have right egg & matching yolks
                (_, "eggwhiteportion", _)                   => !byEntity.Controls.Sprint && IsCrackableEggType(eggType),                            // egg white in container, partial cracking and right egg
                (_, "eggyolkportion", var yolk)             => eggType == "eggyolk" && yolk == eggVariant,                                          // yolks - egg variant must match
                _                                           => false
            };
```

### Version B:
```cs
            var liquidOptions = GetLiquidOptions(blockAccessor, blockSel);

            string eggType = slot.Itemstack.Collectible.FirstCodePart(0);      //grabs currently held item's code
            string eggVariant = slot.Itemstack.Collectible.FirstCodePart(1);   //grabs 1st variant in currently held item

            var existingLiquidState = liquidOptions switch {
                null => null,
                var lo => new {
                    IsEmpty = lo.Value.ExistingStack == null,
                    Type = lo.Value.ExistingStack?.Collectible?.FirstCodePart(0),                       //grabs target liquid's code
                    Variant = lo.Value.ExistingStack?.Collectible?.FirstCodePart(1),                    //grabs target liquid's 1st variant 
                }
            };

            bool canCrack = existingLiquidState switch {
                null                                                                    => false,                                                       // no liquid options for target
                { IsEmpty: true }                                                       => true,                                                        // empty target stack, always succeed
                { Type: var t, Variant: var v } when v == eggVariant => t switch {                                                                      // we have the same variant egg, check on liquid type
                    "eggyolkfullportion"                                                => byEntity.Controls.Sprint && IsCrackableEggType(eggType),     // liquid egg, need to be full cracking
                    "eggwhiteportion"                                                   => !byEntity.Controls.Sprint && IsCrackableEggType(eggType),    // egg white, must *not* be full cracking
                    "eggyolkportion"                                                    => eggType == "eggYolk",                                        // yolk, must start with a yolk item                    
                    _                                                                   => false
                },
                _                                                                       => false
            };
```

B is "simpler" to follow in that it has no discards (`_`) for any of the options, but has a short preamble. But then it ensures chicken & pheasant egg whites dont get mixed together to "chicken egg whites" (although the current functionality on egg white mixability could easily be maintained).

#### Additional remarks:
I _think_ there is some potential danger with the manual checks of `byEntity.Controls.Sprint` in the interaction start/stop methods. This _could_ run out of sync with the assigned hotkeys for the world interactions if someone does not pay too much attention on editing.

Some solution of 
```cs
class EggCrackingType {
    EggCrackingType (EnumEntityAction) {...}
    EggCrackingType (params EnumEntityAction[]) {...}   
    bool IsSatisfied(EntitySelection) {...}
    
    static EggCrackingType Full { get; } = new EggCrackingType(EnumEntityAction.Sprint, EnumEntityAction.Sneak);
    static EggCrackingType Partial { get; } = new EggCrackingType(EnumEntityAction.Sneak);
}

public class EggCrackingWorldInteractionExtender {
  public WorldInteraction WithhotKeysFrom(this WorldInteraction, EggCrackingType ) {...}
}


ItemEggCrack.OnLoaded(ICoreAPI api) {
     ....
     return new WorldInteraction[]
       {
         new WorldInteraction {
           ...
         }.WithHotKeysFrom(EggCrackingType.Full),
         new WorldInteraction {
           ...
         }.WithHotKeysFrom(EggCrackingType.Partial),         
       };
     }
   }
}
```
and then replacing all the `byEntity.Controls.[Foo]` checks with `IsSatisfied` on the appropriate `EggCrackingType` would prevent this from ever going out of sync.

DoctorVanGogh commented:
# Changes:

- Can crack eggs into saucepans or _open_ cauldrons. Can in fact crack eggs into *any* liquid sink. Yes, that includes the boiler & condenser.... 🤡 
- Can aim select which bowl/jug/bottle to crack into if targeting ground storage. *
![image](https://user-images.githubusercontent.com/324067/217673534-325c178e-b1c8-410c-ac70-1e1426783d04.png)
- Can squeeze **Honey** into saucepans or _open_ cauldrons. Can in fact squeeze honey into _any_ liquid sink. 
![image](https://user-images.githubusercontent.com/324067/217673673-089fbcbc-b29a-4b92-9cea-94ba4720456f.png)
- Can aim select which bowl/jug/bottle to squeeze honey into if targeting ground storage. *

---
*: No, vanilla doesn't do that. Vanilla picks top left corner.... and that's it.

# Demo

### Fun with cauldrons
[![Cauldron demo](https://img.youtube.com/vi/ww-4cxuSla8/maxresdefault.jpg)](https://www.youtube.com/watch?v=ww-4cxuSla8)

### Honey squeezing
[![Honey squeezing](https://img.youtube.com/vi/pQLSyA6Ex7U/maxresdefault.jpg)](https://www.youtube.com/watch?v=pQLSyA6Ex7U)

### Egg cracking
[![Honey squeezing](https://img.youtube.com/vi/PffG8rt7O30/maxresdefault.jpg)](https://www.youtube.com/watch?v=PffG8rt7O30)

l33tmaan commented:
Oh goodness, this is all fantastic. Thanks for smoothing out some of my jank. :sweat: 
I can tell you're using the shaders mod!

l33tmaan commented:
Wait, it says there are conflicts with Util/Patches.cs?

DoctorVanGogh commented:
no there aren't :D

DoctorVanGogh commented:
Okay... mucked around a tad more in there.

- Moved things around to sensible locations, "fixed" a bug that would have allowed to squeeze into cauldrons only when closed if they ever get ground-storable and blocked cracking/squeezing on liquid sinks that say they cannot receive liquids?
- Extended the "pick suitable ground storage slot" to include a check for fullness before locking in on that slot.
- **Also** made the egg cracking particles a bit more dynamic. Before particles always spawned at position X (or if the target was a bucket at position Y). Changed that - at least partially successfully - to appear at the "upper edge" of the targeted container facing the current aiming point. What I've got so far works great for direct containers (bucket/cauldron/etc) and not so much for round stored containers.

Observations:

| Item | Particles |
|:---|:---|
| Bucket |  Particles move nicely along the edge depending on the aimpoint |
| Cauldron (small) | Things work like the bucket, but they are slightly too high. Which is expected, since the collision box for small cauldrons is too high 😁 |
| Cauldron (large) | Things work like the bucket... but they feel just _slightly_ too low. Might be the inverse of the small cauldron |
| Ground stored items | Particle location dithers around anemically in the middle-ish |

My current approach is to extract the collision boxes for the target (itemslot's) block, calculate the intersection of an ellipses covering the top plane of the topmost collision box of that block with a line going from the center of that ellipses to the aimpoint and then spawn the particles there at the height of the the top collision box plane. 
![image](https://user-images.githubusercontent.com/324067/217876364-367aa0e5-adca-459a-8301-7c42d43a21b4.png)
It works... until I'm in sub coordinate systems 😢 

I think I _know_ what I'm doing wrong, but I also know I don't know enough about transformations in general and VS' implementation to not make a mess out of this. It's probably something like making some coordinates relative to the sub-slot before doing some calculations, then correctly adding things back on again after.

I'd prefer leaving _that_ problem to someone more capable. So this is probably "good enough" for me.
